File: October/3/reversell2.js
// LeetCode 92: Reverse Linked List II
// Explanation:
// 1. Use dummy node to handle head reversals.
// 2. Traverse to node before 'left'.
// 3. Reverse sublist using head-insertion method.
// 4. Reconnect reversed part back.
// Time Complexity: O(n)
// Space Complexity: O(1)

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
var reverseBetween = function(head, left, right) {
    if (!head || left === right) return head;
    let dummy = new ListNode(0, head);
    let prev = dummy;

    // Move prev to node before 'left'
    for (let i = 0; i < left - 1; i++) {
        prev = prev.next;
    }

    let curr = prev.next;
    for (let i = 0; i < right - left; i++) {
        let temp = curr.next;
        curr.next = temp.next;
        temp.next = prev.next;
        prev.next = temp;
    }

    return dummy.next;
};

// Helper functions
function buildList(arr) {
    let dummy = new ListNode(0), curr = dummy;
    for (let v of arr) {
        curr.next = new ListNode(v);
        curr = curr.next;
    }
    return dummy.next;
}
function toArray(head) {
    let res = [];
    while (head) {
        res.push(head.val);
        head = head.next;
    }
    return res;
}

// Example 1
let head1 = buildList([1,2,3,4,5]);
let res1 = reverseBetween(head1, 2, 4);
console.log(toArray(res1)); // [1,4,3,2,5]

// Example 2
let head2 = buildList([5]);
let res2 = reverseBetween(head2, 1, 1);
console.log(toArray(res2)); // [5]

File: 2025/October/3/reversell2.py
# LeetCode 92: Reverse Linked List II
# Explanation:
# 1. We need to reverse a linked list between positions left and right (1-indexed).
# 2. Use a dummy node to simplify edge cases (reversing from head).
# 3. Traverse to node before 'left', then reverse sublist until 'right'.
# 4. Reconnect reversed sublist back to main list.
# Time Complexity: O(n)
# Space Complexity: O(1)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        if not head or left == right:
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        # Step 1: Move prev to node before 'left'
        for _ in range(left - 1):
            prev = prev.next

        # Step 2: Reverse sublist from left to right
        curr = prev.next
        nxt = None
        for _ in range(right - left):
            nxt = curr.next
            curr.next = nxt.next
            nxt.next = prev.next
            prev.next = nxt

        return dummy.next

# Helper functions
def build_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next

def to_array(head):
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res

# Example 1
head = build_list([1,2,3,4,5])
res = Solution().reverseBetween(head, 2, 4)
print(to_array(res))  # [1,4,3,2,5]

# Example 2
head = build_list([5])
res = Solution().reverseBetween(head, 1, 1)
print(to_array(res))  # [5]
